### ISSUE TITLE
Implement Data Change Notifications Callback

### ISSUE THREAD

**Issue by allanrsmith** [aspnet/EntityFrameworkCore#13827]
### Data Change Notifications Callback
Unable to register a callback on the SqliteConnection object for updates to specified tables

### Functional impact
With separate App/DbContext instances I would like callbacks to be called in a second instance when one instance updates a table row.  This will allow for performant responsive applications.

### Minimal repro steps
Functionality is missing

### Expected result
Register a callback on the SqliteConnection for a specific table/set of tables, perhaps for types of table changes, to allow my code to react to table updates.

### Actual result
No functionality exists to support this

### Further technical details
https://sqlite.org/c3ref/update_hook.html

CC: @bricelam

**Comment by 0xced** [comment]
Issue #441 is about renaming the OnModelCreating argument from *builder* to *modelBuilder*. How is this related to data change notifications callback? Was the issues migrated from another repository without fixing the issue numbers? I am confused. 😕

**Comment by bricelam** [comment]
Yep, migrated. See https://github.com/aspnet/Microsoft.Data.Sqlite/issues/441

**Comment by MikeK-10** [comment]
Looking for documentation or examples for SqliteDataChangeEventHandler if possible?

**Comment by bricelam** [comment]
No docs. This feature isn't implemented yet.

Personally, I think this feature is kinda useless since `sqlite3_update_hook` only fires events for data changed by the *current connection*. In other words, you won't get notifications for data changed by other connections (or applications).

**Comment by Ciantic** [comment]
I was experimenting with this, sqlite3_update_hook seems to be firing on changes happening per connection. It makes it rather useless for multi process subscribers/events.

It might still be useful for someone, it can be of course used even if the patch never lands, just use it directly from SQLitePCL.raw:

```C#
    public static void ListenSqlite(DbContext context)
    {
        context.Database.OpenConnection();
        var c = context.Database.GetDbConnection() as SqliteConnection;
        SQLitePCL.raw.sqlite3_update_hook(c.Handle, test_func, new object());
    }

    static readonly delegate_update test_func = (db, wint, char1, char2, sqlitei64) =>
    {
        Console.WriteLine("Yeah?");
    };
```

In CSProj:
```xml
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="6.0.0" />
    <PackageReference Include="Microsoft.Data.Sqlite" Version="6.0.0" />
    <PackageReference Include="SQLitePCLRaw.core" Version="2.0.7" />
```

**Comment by jtheisen** [comment]
@Ciantic Can somebody else confirm this? It would make the Sqlite documentation irresponsibly misleading as it says nothing about this restriction and it would make this feature indeed almost entirely useless.

Each connection itself obviously already knows what it does.

If that's true they need to update their docs.

**Comment by jtheisen** [comment]
I just got [confirmation](https://sqlite.org/forum/forumpost/2a3c3c9d97) that @Ciantic is correct.

**Comment by bricelam** [comment]
> ...it would make this feature indeed almost entirely useless.
>
> Each connection itself obviously already knows what it does.

💯 Completely agree--that's why this really hasn't been a priority for the team. Just track your own changes in app code.

**Comment by bricelam** [comment]
Most people are probably looking for an event that will fire when data is changed by another process. This is not that.

**Comment by westfin** [comment]
> Most people are probably looking for an event that will fire when data is changed by another process. This is not that.

Hm, Hi @bricelam but description from issue exactly about that:

`With separate App/DbContext instances I would like callbacks to be called in a second instance when one instance updates a table row. This will allow for performant responsive applications.`

Can you please clarify. And do you know any other way to notify application about changes in database?

**Comment by ryanmsnyder** [comment]
Is there a workaround to listen for changes made by other processes?

**Comment by sgf** [comment]
This event can be registered on each connection and directed to the same delegate. In this way, at least within the process, we get event notifications for all connections.

**Comment by bradleyables22** [comment]
I just created a ticket to do this and was notified of this thread. I have a working solution in my own personal project that I thought would be a great addition. 

In my project it when the connection is opened you register a subscription to the notification, and it pushes the changes to a channel.

https://github.com/bradleyables22/LiteGate/blob/master/Server/Interaction/SqliteHooks.cs
https://github.com/bradleyables22/LiteGate/blob/master/Server/Interaction/DatabaseGateManager.cs

https://github.com/dotnet/efcore/issues/36831

**Comment by roji** [comment]
@bradleyables22 see the discussion above on why sqlite3_update_hook is not helpful as a general data change notifications callback - it apparently only reports on changes done by the the connection it's on, not by e.g. external processes. Though as suggested above, I can see that some applications could benefit from in-process notifications only (which is what this would do).

But for that to work, users must be sure to always have the callback registered on each and every connection. As an API, I believe that would make sense on SqliteDataSource (see #30991), which would allow users to configure the callback once on SqliteDataSourceBuilder, at which point all connections handed out by the resulting SqliteDataSource would automatically have it. So I'd argue that we should implement #30991 before doing this.
